pkg/agent/openflow/pipeline.go

```diff
@@ -624,6 +632,112 @@ func (c *client) localProbeFlow(localGatewayIP net.IP, category cookie.Category)
+			Action().ResubmitToTable(conntrackTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Enforce IP packet into the conntrack zone with SNAT. If the connection is SNATed, the reply packet should use
+		// Pod IP as the destination, and then resubmit to conntrackStateTable.
+		c.pipeline[conntrackTable].BuildFlow(priorityNormal).MatchProtocol(binding.ProtocolIP).
+			Action().CT(false, conntrackStateTable, ctZone).NAT().CTDone().
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Rewrite dMAC with the global vMAC if the packet is a reply to the Pod from the external address
+		c.pipeline[conntrackStateTable].BuildFlow(priorityHigh).
+			MatchProtocol(binding.ProtocolIP).
+			MatchCTStateNew(false).MatchCTStateTrk(true).
+			MatchCTMark(snatCTMark).
+			MatchRegRange(int(marksReg), markTrafficFromUplink, binding.Range{0, 15}).
+			Action().LoadRange(binding.NxmFieldDstMAC, vMACInt, binding.Range{0, 47}).
+			Action().ResubmitToTable(dnatTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Resubmit the packet sent from local Pod to the external IP address to dnatTable.
+		c.pipeline[conntrackStateTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchCTStateNew(false).MatchCTStateTrk(true).
+			MatchCTMark(snatCTMark).
+			Action().ResubmitToTable(dnatTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Output the non-SNAT packet to the bridge interface directly if it is received from the uplink interface.
+		c.pipeline[conntrackStateTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchInPort(uplinkOfport).
+			Action().Output(bridgeLocalPort).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Resubmit the packet to L2ForwardingOutput table after the packet is SNATed. The "SNAT" packet has these
+		// characters: 1) the ct_state is "+new+trk", 2) reg1[17] is set as 1; 3) Node IP is used as the target
+		// source IP in NAT action, 4) ct_mark is set with 0x40 in the conn_track context.
+		c.pipeline[conntrackCommitTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchCTStateNew(true).MatchCTStateTrk(true).
+			MatchRegRange(int(marksReg), snatRequiredMark, snatMarkRange).
+			Action().CT(true, l2ForwardingOutTable, ctZone).
+			SNAT(snatIPRange, nil).
+			LoadToMark(snatCTMark).CTDone().
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+	}
+	return flows
+}
+
+func (c *client) l3ToExternalFlows(nodeIP net.IP, localSubnet net.IPNet, outputPort int, category cookie.Category) []binding.Flow {
+	flows := []binding.Flow{
+		// Resubmit the packet to L2ForwardingCalc table if it is communicating to a Service.
+		c.pipeline[l3ForwardingTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
```

Comment:
How we match the first packet of the connection (I think gatewayCTMark is not set yet right?)? Why we do not match ServivceCIDR here?

Reply:
Ok. I now get the first packet will bypass table 70.

Reply:
Wait a min - here if Service traffic already bypass table 70, why we need this flow?

Reply:
This flow is for the subsequent packets of Service traffic. If we use the kube-proxy, it should be the reply packtes for Service traffic (for it is sent from local Pod, and has the gatewayCTMark. gatewayCTMark is committed when the first packte is completed DNAT by kube-proxy, and re-enter OVS pipeline from gw0). We only bypass table70 for the packet which is using Service IP directly.

Reply:
Ok. I assume it is for the return traffic when server and client are on the same Node?

Reply:
But it also catches the return traffic of other connections from gw0 to local Pods, like gw0 -> Pod or Node IP -> Pod?

Reply:
Probably also "another host IP -> a remote Pod", and in this case we should remove matching "markTrafficFromLocal"?

Reply:
> But it also catches the return traffic of other connections from gw0 to local Pods, like gw0 -> Pod or Node IP -> Pod?

Yes, for gw0->Pod traffic, this flow also catches the return traffic.

> Probably also "another host IP -> a remote Pod", and in this case we should remove matching "markTrafficFromLocal"?

I didn't get what you mean about "another host IP". If you mean another node accesses the NodePort Service on this Node, and then targeting at a remote Pod, I don't think this flow could catch the packet. In that case, the return packet should use antrea-gw0's IP as dst on Windows, and the packet should match the generic flow which is using the gw0's address as nw_dst. So does the traffic for current Node to a remote Pod.

Reply:
I meant if the host has more than one IP, then the traffic from another IP (not the Node IP on bridge) to a local Pod will also go gw0 , and should be caught by this flow.
gw0 IP -> Pod is caught by line 716 already.

Reply:
Right, if the Node has more than one IP, and the traffic is from another IP of the host, we should expect the packet could enter OVS pipeline from gw0 (we should enable IP Forwarding on that interface to ensure the packet could be forwarded to gw0 first). And in this case, this flow should catch the return packet. Besides, we could remove the limitation on the source of the packet is local Pod to support the return packet from  a remote Pod.

Reply:
After more thinking, I think typically host IP -> remote Pod should not work, as the return traffic will not go through the tunnel unless we add static routes on the remote Node. So, probably no need to support that case.